**This is an simple spam classifier using methods such as lemmatization and stemming and using bag of words and TF-idf method to predict whether an message is a spam or not and using k-fold method to get the average accuracy of the models.**

In [ ]:
import pandas as pd

msg = pd.read_csv("SMSSpamCollection",sep="\t",names=["type","message"])

In [ ]:
msg.head()

type                                            message
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

**Just using stemming and bag of words model and calculate the accuracy of the model by using naive bayes method for spam classification.**

In [ ]:
import pandas as pd

msg = pd.read_csv("SMSSpamCollection",sep="\t",names=["type","message"])
import re
import nltk
from nltk.stem.porter import PorterStemmer
from nltk.corpus import stopwords
nltk.download('stopwords')

ps = PorterStemmer()
fin_msg = []

for i in range(0,len(msg)):
  review = re.sub("[^a-zA-Z]"," ",msg["message"][i])
  review = review.lower()
  review = review.split()
  review = [ps.stem(word) for word in review if not word in stopwords.words("english")]
  review = " ".join(review)
  fin_msg.append(review)

from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features=2500)
X = cv.fit_transform(fin_msg).toarray()
y=pd.get_dummies(msg["type"])
y=y.iloc[:,1].values

from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import StratifiedKFold
folds = StratifiedKFold(n_splits=3)
scores=[]
mnb = MultinomialNB()
for train_index, test_index in folds.split(X,y):
  X_train, X_test, y_train, y_test = X[train_index], X[test_index],y[train_index], y[test_index]
  mnb.fit(X_train,y_train)
  scores.append(mnb.score(X_test,y_test))
print(sum(scores)/len(scores))

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
0.9820533792268473


**By using k fold method for an 3 splits we are gettin an average score of 98%**

**Now using Lemmatization and bag of words method to improve the model.**

In [ ]:
import pandas as pd

msg = pd.read_csv("SMSSpamCollection",sep="\t",names=["type","message"])
import re
import nltk
from nltk.stem import WordNetLemmatizer
from nltk.corpus import stopwords
nltk.download('stopwords')
nltk.download("wordnet")

wnl = WordNetLemmatizer()
fin_msg = []

for i in range(0,len(msg)):
  review = re.sub("[^a-zA-Z]"," ",msg["message"][i])
  review = review.lower()
  review = review.split()
  review = [wnl.lemmatize(word) for word in review if not word in stopwords.words("english")]
  review = " ".join(review)
  fin_msg.append(review)

from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features=2500)
X = cv.fit_transform(fin_msg).toarray()
y=pd.get_dummies(msg["type"])
y=y.iloc[:,1].values

from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import StratifiedKFold
folds = StratifiedKFold(n_splits=3)
scores=[]
mnb = MultinomialNB()
for train_index, test_index in folds.split(X,y):
  X_train, X_test, y_train, y_test = X[train_index], X[test_index],y[train_index], y[test_index]
  mnb.fit(X_train,y_train)
  scores.append(mnb.score(X_test,y_test))
print(sum(scores)/len(scores))

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
0.9782842449336377


**By using k fold method for an 3 splits we are gettin an average score of 97.8%**


**By using Lemmatization and TF-idf method due to the disadvantages in bag of words model**

In [ ]:
import pandas as pd

msg = pd.read_csv("SMSSpamCollection",sep="\t",names=["type","message"])
import re
import nltk
from nltk.stem import WordNetLemmatizer
from nltk.corpus import stopwords
nltk.download('stopwords')
nltk.download("wordnet")

wnl = WordNetLemmatizer()
fin_msg = []

for i in range(0,len(msg)):
  review = re.sub("[^a-zA-Z]"," ",msg["message"][i])
  review = review.lower()
  review = review.split()
  review = [wnl.lemmatize(word) for word in review if not word in stopwords.words("english")]
  review = " ".join(review)
  fin_msg.append(review)

from sklearn.feature_extraction.text import TfidfVectorizer
Tfv = TfidfVectorizer(max_features=2500)
X = Tfv.fit_transform(fin_msg).toarray()
y=pd.get_dummies(msg["type"])
y=y.iloc[:,1].values

from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import StratifiedKFold
folds = StratifiedKFold(n_splits=3)
scores=[]
mnb = MultinomialNB()
for train_index, test_index in folds.split(X,y):
  X_train, X_test, y_train, y_test = X[train_index], X[test_index],y[train_index], y[test_index]
  mnb.fit(X_train,y_train)
  scores.append(mnb.score(X_test,y_test))
print(sum(scores)/len(scores))

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
0.9772073356198939


**Interestingly the bag of words method worked better than tf-idf method in the case of lemmatization**

**By using stemming and TF-idf method due to the disadvantages in bag of words model**

In [ ]:
import pandas as pd

msg = pd.read_csv("SMSSpamCollection",sep="\t",names=["type","message"])
import re
import nltk
from nltk.stem.porter import PorterStemmer
from nltk.corpus import stopwords
nltk.download('stopwords')
nltk.download("wordnet")

ps = PorterStemmer()
fin_msg = []

for i in range(0,len(msg)):
  review = re.sub("[^a-zA-Z]"," ",msg["message"][i])
  review = review.lower()
  review = review.split()
  review = [ps.stem(word) for word in review if not word in stopwords.words("english")]
  review = " ".join(review)
  fin_msg.append(review)

from sklearn.feature_extraction.text import TfidfVectorizer
Tfv = TfidfVectorizer(max_features=2500)
X = Tfv.fit_transform(fin_msg).toarray()
y=pd.get_dummies(msg["type"])
y=y.iloc[:,1].values

from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import StratifiedKFold
folds = StratifiedKFold(n_splits=3)
scores=[]
mnb = MultinomialNB()
for train_index, test_index in folds.split(X,y):
  X_train, X_test, y_train, y_test = X[train_index], X[test_index],y[train_index], y[test_index]
  mnb.fit(X_train,y_train)
  scores.append(mnb.score(X_test,y_test))
print(sum(scores)/len(scores))

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
0.9761303296963613
